### DISCUSSION TITLE
Question about OpenGrok performance

### DISCUSSION THREAD

**Discussion by a-niec** [oracle/opengrok#4089] [Q&A]
Hi, 
I would like to share some information about OpenGrok performance on our environment.

Node description:
- CPU: Threadripper 3990X (64c 128t)
- RAM: 256GB
- Storage: XFS file system on HDD Raid ~900 MB/sec read/write
- OS: Ubuntu 20.04
- Source base: 360GB in 2.5M files

OpenGrok parameters: 
- opengrok docker image – v 1.7.37
- java heap size 32G (-J=-Xmx32g)
- history disabled
- repo disabled
- lucene memory 256M

Typically indexing source base take **~2h50m**. 

But I noticed that **CPU utilization is really small**. I’ve been monitoring indexing in htop and dstat, and it shows 300-800% CPU usage so only small portion of Threadripper power (5%) is used. Clearly there is some bottleneck here.

I have question for you guys – what were the largest CPU load you saw when indexing?
Maybe you have some tips how to unlock full potential?

Best Regards,
Adrian

**Comment by vladak** [comment]
Hi Adrian ! Is this initial or incremental reindex ? If the former, then 3 hours for initial index of 360 GB of data/code is not so bad at all I'd say.

Also, I wonder how exactly you disabled the history. There is no tunable in the official image that would allow this.

Unless the container is [configured with CPU limits](https://docs.docker.com/config/containers/resource_constraints/#configure-the-default-cfs-scheduler), it should use all CPU power available. The indexer in the container is run via `opengrok-sync`, specifically: https://github.com/oracle/opengrok/blob/8a7aa08e2f11ac1acfba209e18a1444600fe18df/docker/sync.yml#L11-L15 (noting that in high parallelism cases the `RLIMIT_NOFILE` resource setting might be to low - maybe it should be dynamic).

This means that there is no thread cap imposed by the tooling. Consequently, the indexer will choose the parallelism level based on the number of available CPUs: https://github.com/oracle/opengrok/blob/8a7aa08e2f11ac1acfba209e18a1444600fe18df/opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/RuntimeEnvironment.java#L1145-L1154

When the indexer is run without history processing, only 2nd phase of the indexing will be executed. This phase is [heavily parallelized](https://github.com/oracle/opengrok/blob/8a7aa08e2f11ac1acfba209e18a1444600fe18df/opengrok-indexer/src/main/java/org/opengrok/indexer/index/IndexDatabase.java#L1654-L1676).

That said, back to the Docker image. When projects are enabled, the `opengrok-reindex-project` tools, which really is just a simple wrapper of the java run of the indexer, will be executed for each project in parallel, like so:
```
opengrok-latest | 2022-11-04 11:01:30,906    DEBUG opengrok_tools.utils.commandsequence | command = 'opengrok-reindex-project --printoutput --jar /opengrok/lib/opengrok.jar -U http://localhost:8080/ -P Tomcat-10 -- -r dirbased -G -m 256 --leadingWildCards on -c /usr/local/bin/ctags -U http://localhost:8080/ -H Tomcat-10'
```
And again, each of these will run with as many threads as there are CPUs in the system, assuming the container has not been configured otherwise previously. These threads from distinct `opengrok-reindex-project` runs (within the same container or multiple containers) will be competing for the CPU. So, even if there is some project that would have disproportionately bigger amount of data to be processed to the other projects, any decrease of CPU usage should not be visible, because it has thread count set to number of CPUs. Of course, assuming the indexer is CPU bound.

One way to check this would be to determine the number of spawned threads and/or ctags processes (there is 1-1 correspondence) by the indexer while it is running. If there are no ctags processed spawned by the indexer process, the indexer has not reached the 2nd phase of the indexing yet.

For example, when I run the indexing on my laptop with 8 "CPUs" (`cat /proc/cpuinfo | grep ^processor | wc -l` says 8 both in the "global zone" and the container) and internal SSD, using the Docker image, once it gets past the history cache creation to the 2nd phase of the indexing, `htop` output looks like this when indexing a source root that has just a clone of the Linux kernel:

![htop-docker-indexer-2nd_phase](https://user-images.githubusercontent.com/2934284/199979042-c1538626-1491-4006-a42a-ba96ae10623f.png)

I confirmed the indexer parallelism level by running `pgrep -f -l '^\/usr\/local\/bin\/ctags' -c` in the container (reported 8 processes).

Anyhow, while this is definitely way above the 5% CPU utilization, it seems there might be some room for improvement, assuming the I/O does not block the indexer too much. Perhaps raising the thread count above the CPU count could help.

**Reply by vladak** [reply]
Also, just to be sure, I ran the same indexer (from scratch) outside of Docker and the CPU load characteristics in the 2nd phase appear to be the same. Tried increasing the number of indexer threads to 16 and also to 32, did not seem to make much difference in terms of CPU load.

**Reply by a-niec** [reply]
>Is this initial or incremental reindex ? If the former, then 3 hours for initial index of 360 GB of data/code is not so bad at all I'd say.

This is initial commit. I know that 3h is not bad for such big code base, but I'm still thinking that such powerful machine could do better.

> Also, I wonder how exactly you disabled the history. There is no tunable in the official image that would allow this.

True - stock /script/starty.py does indexing with history by default. I've extracted it from image, override it (removed -H) and mounted to container. Also I'm passing '--historyBased off --remote off' to indexer.

> Unless the container is [configured with CPU limits](https://docs.docker.com/config/containers/resource_constraints/#configure-the-default-cfs-scheduler), it should use all CPU power available. The indexer in the container is run via opengrok-sync, specifically...

There are no limitations for container - it uses all cores, and have decent limits:
```bash
ulimit -Ha
core file size          (blocks, -c) unlimited
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 1030939
max locked memory       (kbytes, -l) 65536
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1048576
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) unlimited
cpu time               (seconds, -t) unlimited
max user processes              (-u) 1030939
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```
I'm not using auto sync (also disabled in start.py) because I want to have exact start and end time in script (pipeline). I'm calling following command to index project.

```bash
opengrok-reindex-project --printoutput -J=-Xmx32g -J=-server -J=-Djava.util.logging.config.file=/opengrok/logging.properties \
--jar /opengrok/lib/opengrok.jar -U http://localhost:8080/${URL_ROOT}/ -P PROJECT01 -- \
--progress -m 256 -c /usr/local/bin/ctags --historyBased off --remote off \
-U http://localhost:8080/${URL_ROOT}/ PROJECT01
```

> This means that there is no thread cap imposed by the tooling. Consequently, the indexer will choose the parallelism level based on the number of available CPUs

True - when indexing there is 128 threads of ctags  and all of them are working (rather slacking),  but on very low load. I've also tried with 64,32 and 16 threads (passed -T # to indexer) with very similar effect. 
Example ot -T 64 :
![htop](https://user-images.githubusercontent.com/115471600/199992839-fc4821bb-1d82-4e36-b1db-28d9f4975ceb.png)

Note that there is one CPU thread running at 100% - it is running like this all the time until threads are set to ~10.

I'm guessing that this may be **main java process** (the one which spawns all ctag threads) and it saturates when there are lot of threads to handle.

So my hypothesis is: **if threads are more than 10 threads - main java process is starving due to  some single-core workload**

**Reply by vladak** [reply]
In my run within IDEA, I noticed this in one of the thread snapshots. There were 30 threads (running with `--threads 32`) parking at this spot:
```
"ForkJoinPool-1-worker-37" #22 daemon prio=5 os_prio=0 cpu=7362,01ms elapsed=223,51s tid=0x00007f7300285800 nid=0x3d33b waiting on condition  [0x00007f72eb2f6000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.16/Native Method)
	- parking to wait for  <merged>(a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.16/LockSupport.java:194)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@11.0.16/CompletableFuture.java:1796)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@11.0.16/ForkJoinPool.java:3118)
	at java.util.concurrent.CompletableFuture.waitingGet(java.base@11.0.16/CompletableFuture.java:1823)
	at java.util.concurrent.CompletableFuture.get(java.base@11.0.16/CompletableFuture.java:1998)
	at org.opengrok.indexer.analysis.plain.PlainAnalyzer.analyze(PlainAnalyzer.java:176)
	at org.opengrok.indexer.analysis.AnalyzerGuru.populateDocument(AnalyzerGuru.java:626)
	at org.opengrok.indexer.index.IndexDatabase.addFile(IndexDatabase.java:1068)
	at org.opengrok.indexer.index.IndexDatabase.lambda$indexParallel$4(IndexDatabase.java:1687)
	at org.opengrok.indexer.index.IndexDatabase$$Lambda$274/0x00007f72eb688968.apply(Unknown Source)
	at java.util.stream.Collectors.lambda$groupingByConcurrent$59(java.base@11.0.16/Collectors.java:1304)
	at java.util.stream.Collectors$$Lambda$276/0x00007f72eb2fed08.accept(java.base@11.0.16/Unknown 
```
which matches https://github.com/oracle/opengrok/blob/8a7aa08e2f11ac1acfba209e18a1444600fe18df/opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/plain/PlainAnalyzer.java#L168-L175

and interestingly the executor is instantiated like this: https://github.com/oracle/opengrok/blob/8a7aa08e2f11ac1acfba209e18a1444600fe18df/opengrok-indexer/src/main/java/org/opengrok/indexer/index/IndexerParallelizer.java#L249-L253

so it looks like there is just a [single thread in the pool](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledThreadPoolExecutor.html#ScheduledThreadPoolExecutor(int,%20java.util.concurrent.ThreadFactory)). Not completely sure if this is the problem, though.

**Reply by a-niec** [reply]
Maybe this will help.
I've enabled custom thread name option in htop.
Indexing looks like this - almost all the time.
**Lucene Merge Thread** 80% time
**OpenGrok-xref-w(riter?)** 100% time
**ForkJoinPool** - depends - 1or2 threads have 100% other ~1-5%


![thread_details](https://user-images.githubusercontent.com/115471600/200007396-9c275c41-2bc1-4b88-a4ba-696ab7a25a18.png)

**Reply by vladak** [reply]
Changing the 1 to `indexingParallelism` in `createLazyXrefWatcherExecutor()` definitely increased the CPU load.

**Reply by vladak** [reply]
The change results in Linux kernel repo indexing time (no -H, with --threads 32 and using 32G heap) decrease from 6:01 minutes to 4:42 minutes. Now I need to double check it does not cause any harm.

**Reply by a-niec** [reply]
Nice :)

so this is only change that result in speedup?

```java
- new ScheduledThreadPoolExecutor(1, 
+ new ScheduledThreadPoolExecutor(indexingParallelism, 
```
If so I'll check this in my environment

**Reply by a-niec** [reply]
Ok, so I've been testing this change with 1 -> indexingParallelism in self-baked docker image.
It is spawning more OpenGrok-xref-writer threads.
 
But unfortunately there is no big improvement in time (2:50 -> 2:35).
About your results - did you clear page cache between tests ? eg. sudo bash -c "echo 3 > /proc/sys/vm/drop_caches" 
Second indexing on cached files is always faster.

I've tested many threads / lucene mem options.

I've even run indexing (smaller set 30GB) of sources on tmpfs and output to tmpfs - to eliminate any io latency.
It was a a bit better ~70% of normal (4x raid) time, but  still - server didn't even heat up ;)

![ramdisk_htop](https://user-images.githubusercontent.com/115471600/200194035-1b0a3965-dc3e-40cf-957b-99643d302fae.png)

![netdata_htop](https://user-images.githubusercontent.com/115471600/200194050-17f97560-aa78-4e22-a8f7-9ebddbdb77ed.png)

Are those Lucene Merging Thread can somehow impact indexing speed? They are always at 100% maybe they are capping indexers?

**Reply by vladak** [reply]
There was another change I did as part of PR #4093: 60b407e217ef9591736dc6b1c984163db61153cb

In my case it made the CPU to be almost fully utilized during indexing. To measure the run times, I ran the indexing multiple times, so in each case the file-system caches should be heated up, although I did not verify.

**Reply by vladak** [reply]
The changes are part of the today released 1.7.38, so you can try official Docker images.

**Reply by a-niec** [reply]
I've re-tested latest image on ramdisk input -> ramdisk output 
This images speaks for themself. 
![final_htop](https://user-images.githubusercontent.com/115471600/200421247-6934c357-2a6a-4760-b784-108a10ed1efa.png)

Tested on 30GB of sources. Finished in 0:03:25 (previously ~0:06:30). Theoretical (ramdisk) throughput is awesome.
Unbelievable what 2 simple variable changes can do to such complex project.

Starting from v1.7.38 OpenGrok is as fast as CPU and storage let him be ;)
